storage/innobase/trx/trx0trx.cc

```diff
@@ -842,9 +842,31 @@ trx_resurrect_insert(
 
 			if (srv_force_recovery == 0) {
 
-				/* XtraBackup should rollback prepared XA
-				transactions */
-				trx->state = TRX_STATE_ACTIVE;
+				/* XtraBackup should rollback internal prepared
+				XA transactions, but leave external prepared
+				XA transactions untouched. */
+				if (trx->xid->get_my_xid()) {
```

Comment:
Thank you for contribution. This change looks reasonable, but it will also do not rollback WSREP XIDs (prefixed with "WSREPXid"). However, Galera Cluster rely on fact that XIDs are rolled back by xtrabackup.

Reply:
Thanks for the information, I'm not familiar with Galera or mysql-wsrep :)
If the XID prefix is fixed for WSREP, I can add another condition check here.
I'll update the PR shortly.

Reply:
I have pushed the updates
1. handle wsrep xid case
2. rebase to latest 2.4